source/common/http/path_utility.cc

```diff
@@ -64,27 +104,76 @@ bool PathUtil::canonicalPath(RequestHeaderMap& headers) {
   if (!query_suffix.empty()) {
     normalized_path.insert(normalized_path.end(), query_suffix.begin(), query_suffix.end());
   }
-  headers.setPath(normalized_path);
-  return true;
+  return normalized_path;
 }
 
-void PathUtil::mergeSlashes(RequestHeaderMap& headers) {
-  ASSERT(headers.Path());
-  const auto original_path = headers.getPathValue();
+absl::optional<std::string> PathTransformer::unescapeSlashes(absl::string_view original_path) {
   // Only operate on path component in URL.
   const absl::string_view::size_type query_start = original_path.find('?');
   const absl::string_view path = original_path.substr(0, query_start);
+  if (path.find('%') == absl::string_view::npos) {
+    return std::string(original_path);
+  }
   const absl::string_view query = absl::ClippedSubstr(original_path, query_start);
-  if (path.find("//") == absl::string_view::npos) {
-    return;
+
+  // TODO(yanavlasov): optimize this by adding case insensitive matcher
+  std::string decoded_path{path};
+  unescapeInPath(decoded_path, "%2F", "/");
+  unescapeInPath(decoded_path, "%2f", "/");
+  unescapeInPath(decoded_path, "%5C", "\\");
+  unescapeInPath(decoded_path, "%5c", "\\");
+  return absl::StrCat(decoded_path, query);
+}
+
+PathTransformer::PathTransformer(const bool should_normalize_path,
+                                 const bool should_merge_slashes) {
+  if (should_normalize_path) {
+    transformations_.emplace_back(PathTransformer::rfcNormalize);
+    normalize_path_actions_.push_back(NormalizePathAction::Continue);
+  }
+  if (should_merge_slashes) {
+    transformations_.emplace_back(PathTransformer::mergeSlashes);
+    normalize_path_actions_.push_back(NormalizePathAction::Continue);
   }
-  const absl::string_view path_prefix = absl::StartsWith(path, "/") ? "/" : absl::string_view();
-  const absl::string_view path_suffix = absl::EndsWith(path, "/") ? "/" : absl::string_view();
-  headers.setPath(absl::StrCat(path_prefix,
-                               absl::StrJoin(absl::StrSplit(path, '/', absl::SkipEmpty()), "/"),
-                               path_suffix, query));
 }
 
+PathTransformer::PathTransformer(
+    envoy::type::http::v3::PathTransformation const& path_transformation) {
+  const auto& operations = path_transformation.operations();
+  std::vector<uint64_t> operation_hashes;
+  for (auto const& operation : operations) {
+    uint64_t operation_hash = MessageUtil::hash(operation);
+    if (find(operation_hashes.begin(), operation_hashes.end(), operation_hash) !=
+        operation_hashes.end()) {
+      // Currently we only have RFC normalization and merge slashes, don't expect duplicates for
```

Comment:
The comment is not correct any more. There are 3 transformations now.

Reply:
Fixed.